model/shift_design_constraints.py

```diff
@@ -20,27 +20,40 @@ def __init__(
         self.long_shifts = shift_sets["long_shifts"]
 
         # Adding constraints
-        self.add_minimum_demand_coverage(var.x)
-        self.add_deviation_from_demand(var.x, var.delta)
+        self.add_minimum_demand_coverage(var.x, var.mu)
+        self.add_maximum_demand_coverage(var.mu)
+        self.add_deviation_from_demand(var.mu, var.delta)
         self.add_mapping_x_to_y(var.x, var.y)
         self.add_short_shift_duration(var.y, var.rho)
         self.add_long_shift_duration(var.y, var.rho)
 
     # Constraint definitions
-    def add_minimum_demand_coverage(self, x):
+    def add_minimum_demand_coverage(self, x, mu):
         self.model.addConstrs(
             (
                 quicksum(x[t_marked, v] for t_marked, v in self.shifts_overlapping_t[t])
-                >= quicksum(self.demand["min"][c, t] for c in self.competencies if self.demand["min"].get((c,t)))
+                == quicksum(self.demand["min"][c, t] for c in self.competencies if self.demand["min"].get((c, t)))
+                + quicksum(mu[c, t] for c in self.competencies)
                 for t in self.time_periods_combined
             ),
             name="minimum_demand_coverage",
         )
 
-    def add_deviation_from_demand(self, x, delta):
+    def add_maximum_demand_coverage(self, mu):
         self.model.addConstrs(
             (
-                quicksum(x[t_marked, v] for t_marked, v in self.shifts_overlapping_t[t])
+                mu[c, t] <= self.demand["max"][c, t] - self.demand["min"][c, t]
+                for c in self.competencies
+                for t in self.time_periods[c]
+            ),
+            name="maximum_demand_coverage"
+        )
+
+    def add_deviation_from_demand(self, mu, delta):
+        self.model.addConstrs(
+            (
+                mu[c, t]
+                + self.demand["min"][c, t]
                 - self.demand["ideal"][c, t]
                 == delta["plus"][c, t] - delta["minus"][c, t]
                 for c in self.competencies
```

Comment:
```python
                mu[t]
                + quicksum(
                + self.demand["min"][c, t]
                - self.demand["ideal"][c, t]
               for c)
                == quicksum(delta["plus"][c, t] - delta["minus"][c, t] for c)
                for c in self.competencies
```

Reply:
Oppdatert

```python
mu[t]
+ quicksum(
    self.demand["min"][c, t]
    - self.demand["ideal"][c, t]
    for c in self.competencies if self.demand["min"].get(c, t)
    )
== 
quicksum(
    delta["plus"][c, t] 
    - delta["minus"][c, t] for c in self.competency if delta["plus"].get(c, t))
    for c in self.competencies
) 
for t in self.combined_time_periods

Reply:
""" python
 mu[t]
                + quicksum(
                + self.demand["min"][c, t]
                - self.demand["ideal"][c, t]
               for c in self.competencies if (c,t) in self.demand["min"])
                == quicksum(delta["plus"][c, t] - delta["minus"][c, t] for c)
                for c in self.competencies if (c,t) in delta["minus])
for t in time_periods_combined)
"""